### DISCUSSION TITLE
Feature request - custom file_name in sndfile Module

### DISCUSSION THREAD

**Discussion by ninp0** [baresip/baresip#1990] [General]
Please implement the means to provide a custom name for a recording when using the sndfile module.  So instead of:

dump-TIMESTAMP-dec.wav
dump-TIMESTAMP-enc.wav

Allow for something along the lines of:

SIP_DEST-TIMESTAMP-dec.wav
SIP_DEST-TIMESTAMP-enc.wav

**Comment by alfredh** [comment]
do you want to contribute a patch with your suggested feature ?

**Comment by ninp0** [comment]
I'm open to doing that, however, I'm not entirely sure what struct out of baresip.h I should be using for the current destination address...would it be the call struct and which function needs to be called?

**Reply by cspiel1** [reply]
You don't need to change a struct.
See how other modules are doing this! E.g. avcodec:
```
	conf_get_str(conf_cur(), "avcodec_h264enc", h264enc, sizeof(h264enc));
```

**Comment by ninp0** [comment]
Yeah that's not what I'm asking...I agree, changing a struct in baresip.h is the wrong approach.  That said, I know one of the structs in baresip.h (perhaps call?) already has the means to retrieve the current sip destination for the current call in progress.  The thought being to pass something like the call struct into https://github.com/baresip/baresip/blob/ca4f5fbaabf2d13c6e1471aa5e97c82bf547e6b7/modules/sndfile/sndfile.c#L82 the *openfile function so that the destination sip could be passed into re_snprintf call at https://github.com/baresip/baresip/blob/ca4f5fbaabf2d13c6e1471aa5e97c82bf547e6b7/modules/sndfile/sndfile.c#L91 ...simply wondering if anyone knows which struct in baresip.h would be passed to the openfile function so I can easily extract the sip dest and update the filename in sndfile.c.

**Reply by cspiel1** [reply]
`sndfile` is an audio filter. It doesn't know anything about calls. But it has a pointer to the associated `audio` object. E.g.
`audio` --> `stream` --> `sdp_media` --> `raddr`

There are get-ter functions in `baresip.h`.

What exactly is the reason for the DEST in the file name?

**Reply by ninp0** [reply]
Simply so I can associate a recording with the sip destination dialed.  In other words, if the following command:
```
/dial 15554443333
```
is executed, the recordings generated by the sndfile module would be 15554443333-TIMESTAMP-dec.wav & 15554443333-TIMESTAMP-enc.wav

A getter function would be perfect - do you know off the cuff which getter function should be called?

**Reply by ninp0** [reply]
Keeping this discussion going - it's starting to look like baresip.h should be changed to include perhaps a _copy_ of the aufilt struct to include the call struct as well?  My initial impression is that it would be easy to have dump-TIMESTAMP>dec.wav & dump-TIMESTAMP-enc.wav to include the destination phone number that is currently being dialed...I'm starting to realize, this probably isn't the case.  @cspiel1 , would it make sense to implement a copy of aufilt to include the call struct?  Any guidance you can provide is very much appreciated.  Thanks in advance.

**Reply by cspiel1** [reply]
@alfredh Should we add a `call` pointer to the `audio` or `stream` struct?

**Reply by alfredh** [reply]
I am currently looking at the auformat module which implements audio source and video source.
it has a concept of "shared" state between audio/video, for example if the backend is an MP4 file
with audio track and video track. It is currently not working so well for multiple sessions,
where the next session will "grab" the previous session. Therefore we need to add something
like a Session identifier (e.g. call-id, cname, etc).

please note that I have added peerconn.c which is WebRTC peer connection, which uses
audio.c and video.c -- So now we have 2 users of audio/video/stream:

1. call.c
2. peerconn.c


You can add a `struct call` pointer to stream if you want, but please keep in mind that this will not work for peerconn.c

One alternative is to add some kind of opaque Session ID string to stream,
that can be used also to solve this problem.

**Reply by cspiel1** [reply]
The call-id seems to be made for this. @ninp0 you are welcome to create a PR. I will be on vacation until 22.08.

**Comment by ninp0** [comment]
I've been spinning my wheels on this for a while now.  The issue I'm running into is properly instantiating the, "const struct call" object within the "openfile" function in:

modules/sndfile/sndfile.c

I assume once I'm able to properly instantiate the call object, I can pass it to one of the  call functions declared in baresip.h:
```
223 const char   *call_id(const struct call *call);
224 const char   *call_peeruri(const struct call *call);
225 const char   *call_peername(const struct call *call);
```


I added audio as parameter to openfile function
```
82 static SNDFILE *openfile(const struct aufilt_prm *prm,
83                                 const struct audio *au,
84                                 bool enc)
85 {
```

I added the three lines to include, as an example the call_peeruri char in the filename:
```
   93         struct stream *strm = audio_strm(au);
   94         const struct call *call = strm->call;
   95         const char *call_peeruri = call_peeruri(call);
   96 
   97         (void)re_snprintf(filename, sizeof(filename),
   98                                 "%s/dump-%s-%H-%s.wav",
   99                                 file_path,
 100                                 call_peeruri,
 101                                 timestamp_print, tm, enc ? "enc" : "dec");
```

However it doesn't appear call is associated with audio and/or stream.

Any help / guidance / hints / clues around instantiating the const struct call object within the openfile function so it can be passed to the call_peeruri function is greatly appreciated.  Thanks.

**Reply by cspiel1** [reply]
Right, a `call` pointer should not be added to stream/audio. But you need only a string.

Look at `struct stream_param`! It already as a `cname`. You may add another `const char *` and maybe can pass this to the filter module.

**Comment by ninp0** [comment]
Unsure if I'm getting any closer here:

I added a "const char *peer_uri" pointer in, "struct stream_param" within include/baresip.h:
```
1362 /** Common parameters for media stream */
1363 struct stream_param {
1364         bool use_rtp;         /**< Enable or disable RTP */
1365         int af;               /**< Wanted address family */
1366         const char *cname;    /**< Canonical name        */
1367         const char *peer_uri; /**< Peer SIP Address      */
1368 };
```

I also updated the following lines in the, "openfile" function within modules/sndfile/sndfile.c:
```
91         const struct stream_param *sp;
92         const char *peer_uri = sp->peer_uri; /* Warning produced at compile time >> sndfile.c:92:21: warning: ‘sp’ is used uninitialized */
93         (void)re_snprintf(filename, sizeof(filename),
94                           "%s/dump-%s-%H-%s.wav",
95                                 file_path,
96                                 peer_uri,
97                           timestamp_print, tm, enc ? "enc" : "dec");
```

Once compiled, dialing any arbitrary phone number in a baresip session produces recordings w/ strange file names such as these:
```
$ ls
'dump-'$'\340'','$'\343''wT'$'\177''-2022-09-21-16-24-26-dec.wav'
'dump-'$'\340'','$'\343''wT'$'\177''-2022-09-21-16-24-26-enc.wav'
```
Thoughts?

**Reply by cspiel1** [reply]
First I would prefer
```
const char *sname;    /**< Stream name    */
```

Second:
Try to figure out how and where this has to be set to reference real data in memory!
The gcc warning is more than I could explain here.

Third:
You are welcome to open a PR! These need not be a "discussion".

**Reply by ninp0** [reply]
Got it working - here's the PR: https://github.com/baresip/baresip/pull/2140
